ERCS/erc-7738.md

```diff
@@ -0,0 +1,152 @@
+description: Permissionless registry to fetch executable scripts for contracts
+author: Victor Zhang (@zhangzhongnan928), James Brown (@JamesSmartCell)
+discussions-to: https://ethereum-magicians.org/t/erc-7738-permissionless-script-registry/20503
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-07-01
+---
+## Abstract
+
+This EIP provides a means to create a standard registry for locating executable scripts associated with the token.
+
+## Motivation
+
+[ERC-5169](./eip-5169.md) (`scriptURI`) provides a client script lookup method for contracts. This requires the contract to have implemented the ERC-5169 interface at the time of construction (or allow an upgrade path).
+
+This proposal outlines a contract that can supply prototype and certified scripts. The contract would be a singleton instance multichain that would be deployed at identical addresses on supported chains.
+
+### Overview
+
+The registry contract will supply a set of URI links for a given contract address. These URI links point to script programs that can be fetched by a wallet, viewer or mini-dapp.
+
+The pointers can be set using a setter in the registry contract.
+
+The scripts provided could be authenticated in various ways:
+
+1. The target contract which the setter specifies implements the `Ownable` interface. Once the script is fetched, the signature can be verified to match the Owner(). In the case of TokenScript this can be checked by a dapp or wallet using the TokenScript SDK, the TokenScript online verification service, or by extracting the signature from the XML, taking a keccak256 of the script and ecrecover the signing key address.
+2. If the contract does not implement Ownable, further steps can be taken:
+ a. The hosting app/wallet can acertain the deployment key using 3rd party API or block explorer. The implementing wallet, dapp or viewer would then check the signature matches this deployment key.
+ b. Signing keys could be pre-authenticated by a hosting app, using an embedded keychain.
+ c. A governance token could allow a script council to authenticate requests to set and validate keys.
+
+If these criteria are not met:
+- For mainnet implementations the implementing wallet should be cautious about using the script - it would be at the app and/or user's discretion.
+- For testnets, it is acceptable to allow the script to function, at the discretion of the wallet provider.
+
+## Specification
+
+The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY” and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+The contract MUST implement the IDecentralisedRegistry interface.
+The contract MUST emit the ScriptUpdate event when the script is updated.
+The contract SHOULD order the scriptURI returned so that the owner script is returned first (in the case of simple implementations the wallet will pick the first scriptURI returned).
```

Comment:
```suggestion
The contract SHOULD order the `scriptURI` returned so that the owner script is returned first (in the case of simple implementations the wallet will pick the first `scriptURI` returned).
```